### ISSUE TITLE
File system recreated upon recoverable mount error, causing data loss

### ISSUE THREAD

**Issue by sgielen** [rancher/k3s#1902]
**Version:**
`k3s version v1.17.4+k3s1 (3eee8ac3)`

**K3s arguments:**
Just `agent`

**Describe the bug**
When a block device cannot be mounted as a filesystem because of an error during mount, the filesystem is wiped and a new filesystem is created. This causes data loss if the mount error was recoverable.

I am using OpenEBS for storage. OpenEBS creates iSCSI targets within the cluster, which offer block devices that are mounted back into the cluster as filesystems. Per Kubernetes specifications, `k3s` performs the creation of these filesystems on the block devices, and takes care of mounting the filesystems into Pods that specify them.

I have such an OpenEBS volume in my cluster, a block device that is mounted as a filesystem. I created it as a 400 GB block device, then resized it to 1500 GB. However, I forgot one of the resizing steps; when the iSCSI target was restarted, the block device was offered as 400 GB again, but the filesystem was 1500 GB.

A Pod was created that mounted the filesystem on the node `karen`. It contains this in its k3s-service.log:

```
I0614 19:07:52.218680    2407 operation_generator.go:1245] Controller attach succeeded for volume "pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7" (UniqueName: "kubernetes.io/iscsi/10.43.50.167:3260:iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7:0") pod "samba-548d78968d-lgpdn" (UID: "860610f1-ce88-44aa-81d2-6d94da06ff28") device path: ""
I0614 19:07:52.284937    2407 operation_generator.go:551] MountVolume.WaitForAttach entering for volume "pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7" (UniqueName: "kubernetes.io/iscsi/10.43.50.167:3260:iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7:0") pod "samba-548d78968d-lgpdn" (UID: "860610f1-ce88-44aa-81d2-6d94da06ff28") DevicePath ""
E0614 19:07:57.652618    2407 mount_linux.go:140] Mount failed: exit status 32
Mounting command: mount
Mounting arguments: -t ext4 -o defaults /dev/disk/by-path/ip-10.43.50.167:3260-iscsi-iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-lun-0 /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.43.50.167:3260-iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-lun-0
Output: mount: /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.43.50.167:3260-iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-lun-0: wrong fs type, bad option, bad superblock on /dev/sdd, missing codepage or helper program, or other error.
I0614 19:07:57.918782    2407 mount_linux.go:322] Disk "/dev/disk/by-path/ip-10.43.50.167:3260-iscsi-iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-lun-0" appears to be unformatted, attempting to format as type: "ext4" with options: [-F -m0 /dev/disk/by-path/ip-10.43.50.167:3260-iscsi-iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-lun-0]
[....]
I0614 19:13:05.766133    2407 mount_linux.go:326] Disk successfully formatted (mkfs): ext4 - /dev/disk/by-path/ip-10.43.50.167:3260-iscsi-iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-lun-0 /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.43.50.167:3260-iqn.2016-09.com.openebs.cstor:pvc-df6179f9-d398-442e-b3fb-aa1f859d43f7-lun-0
```

The `dmesg` contains:

```
[Jun14 19:07] iscsi: registered transport (tcp)
[...]
[  +0.079672] sd 3:0:0:0: [sdd] Attached SCSI disk
[...]
[  +0.843691] EXT4-fs (sdd): bad geometry: block count 393216000 exceeds size of device (104857600 blocks)
```

**To Reproduce**
I think any mount error, on any block-device backing store would reproduce this. But, specifically:
- Set up an iSCSI target limited to 10 GB. (see [k8s volume docs](https://kubernetes.io/docs/concepts/storage/volumes/#iscsi) / [k8s volume examples](https://github.com/kubernetes/examples/tree/master/volumes/iscsi).)
- Start a Pod that mounts the PV. It's initialized to 10 GB. Stop the Pod.
- Restart the iSCSI target limited to 5 GB.
- Start the Pod again.
- Observe k3s logs and dmesg: the filesystem is rewritten on the PV, because the mount fails with "bad geometry: block count .. exceeds size of device..".

**Expected behavior**
k3s should be very careful to create a new filesystem on the device, as this is something that easily causes data loss. If the device looks like it contains an existing filesystem, the pod should fail to start, instead of assuming that the data can be wiped. For example, it should only create a filesystem if the first kilobyte contains only zeroes.

This way, the error can be investigated and solved. If the user wants the filesystem to be recreated (which is rarely the case and indicates critical failure of storage), they can do so themselves, or create a new PV to replace the old one.

**Actual behavior**
Any mount error will cause filesystem to be recreated, causing data loss of everything that was on there.

**Comment by sgielen** [comment]
Code actually seems to come from the Kubernetes upstream: https://github.com/kubernetes-csi/driver-registrar/blob/master/vendor/k8s.io/kubernetes/pkg/util/mount/mount_linux.go#L489

**Comment by brandond** [comment]
This was fixed in k3s 1.18. I'm on mobile so I don't have the issue number handy.

**Comment by sgielen** [comment]
> I believe this was fixed in k3s 1.18.2 - see #1682

Good to hear – I will refine the reproduction steps, so I can try to make this happen one more time; then, I will upgrade to k3s 1.18.2 and try the steps again to confirm it has been fixed. Thanks!

**Comment by sgielen** [comment]
I have some reproduction steps. I'll upgrade to 1.18.2 to try them again and see if it's indeed resolved :-)

### Start a pod & create a 10GB filesystem
$ kubectl run -ti --image ubuntu:bionic tgt bash

**Comment by sgielen** [comment]
Unfortunately, there's no release of k3os with k3s 1.18+ yet, so I upgraded my k3os cluster using these steps:

1. SSH to your master (server) node. Or the first one, if you have multiple.
1. `wget https://github.com/rancher/k3s/releases/download/v1.18.3%2Bk3s1/k3s-arm64`
1. `sudo mount -o remount,rw /k3os/system`
1. `sudo mkdir /k3os/system/k3s/v1.18.3+k3s1`
1. `sudo mv k3s-arm64 /k3os/system/k3s/v1.18.3+k3s1/k3s`
1. `sudo ln -s v1.18.3+k3s1 /k3os/system/k3s/current`
1. `sudo reboot`
1. Repeat these steps on all other master node(s) if you have multiple.
1. Repeat these steps on all worker (agent) nodes.

**Comment by sgielen** [comment]
After the upgrade was completed, I tried it again this morning, and it seems the issue has indeed been resolved in k3s 1.18+.

After the `wipe-pod-test` is started, this is visible in its events (`kubectl describe pod`):

```
  Warning  FailedMount             77s (x9 over 4m39s)  kubelet, karen           MountVolume.MountDevice failed for volume "testvol" : mount failed: exit status 32
Mounting command: mount
Mounting arguments: -t ext4 -o defaults /dev/disk/by-path/ip-10.42.2.21:3260-iscsi-iqn.2020-06.k3s-test:testvol-lun-0 /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0
Output: mount: /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0: wrong fs type, bad option, bad superblock on /dev/sdc, missing codepage or helper program, or other error.
  Warning  FailedMount  47s (x2 over 3m3s)  kubelet, karen  Unable to attach or mount volumes: unmounted volumes=[testvol], unattached volumes=[testvol default-token-gd4nm]: timed out waiting for the condition
```

This is visible in the `k3s-service.log`:

```
I0616 07:31:08.481068    2391 reconciler.go:224] operationExecutor.VerifyControllerAttachedVolume started for volume "testvol" (UniqueName: "kubernetes.io/iscsi/10.42.2.21:3260:iqn.2020-06.k3s-test:testvol:0") pod "wipe-pod-test" (UID: "81fd0125-8d04-4dc4-951f-9c199db3f76f") 
[...]
I0616 07:31:24.187536    2391 operation_generator.go:1332] Controller attach succeeded for volume "testvol" (UniqueName: "kubernetes.io/iscsi/10.42.2.21:3260:iqn.2020-06.k3s-test:testvol:0") pod "wipe-pod-test" (UID: "81fd0125-8d04-4dc4-951f-9c199db3f76f") device path: ""
I0616 07:31:24.263471    2391 operation_generator.go:558] MountVolume.WaitForAttach entering for volume "testvol" (UniqueName: "kubernetes.io/iscsi/10.42.2.21:3260:iqn.2020-06.k3s-test:testvol:0") pod "wipe-pod-test" (UID: "81fd0125-8d04-4dc4-951f-9c199db3f76f") DevicePath ""
Mounting arguments: -t ext4 -o defaults /dev/disk/by-path/ip-10.42.2.21:3260-iscsi-iqn.2020-06.k3s-test:testvol-lun-0 /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0
Output: mount: /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0: wrong fs type, bad option, bad superblock on /dev/sdc, missing codepage or helper program, or other error.
E0616 07:31:35.598986    2391 iscsi_util.go:510] iscsi: failed to mount iscsi volume /dev/disk/by-path/ip-10.42.2.21:3260-iscsi-iqn.2020-06.k3s-test:testvol-lun-0 [ext4] to /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0, error mount failed: exit status 32
Mounting arguments: -t ext4 -o defaults /dev/disk/by-path/ip-10.42.2.21:3260-iscsi-iqn.2020-06.k3s-test:testvol-lun-0 /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0
Output: mount: /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0: wrong fs type, bad option, bad superblock on /dev/sdc, missing codepage or helper program, or other error.
I0616 07:31:35.599270    2391 operation_generator.go:567] MountVolume.WaitForAttach succeeded for volume "testvol" (UniqueName: "kubernetes.io/iscsi/10.42.2.21:3260:iqn.2020-06.k3s-test:testvol:0") pod "wipe-pod-test" (UID: "81fd0125-8d04-4dc4-951f-9c199db3f76f") DevicePath "/dev/disk/by-path/ip-10.42.2.21:3260-iscsi-iqn.2020-06.k3s-test:testvol-lun-0"
[...]
I0616 07:34:57.490862    2391 operation_generator.go:567] MountVolume.WaitForAttach succeeded for volume "testvol" (UniqueName: "kubernetes.io/iscsi/10.42.2.21:3260:iqn.2020-06.k3s-test:testvol:0") pod "wipe-pod-test" (UID: "81fd0125-8d04-4dc4-951f-9c199db3f76f") DevicePath "/dev/disk/by-path/ip-10.42.2.21:3260-iscsi-iqn.2020-06.k3s-test:testvol-lun-0"
Mounting arguments: -t ext4 -o defaults /dev/disk/by-path/ip-10.42.2.21:3260-iscsi-iqn.2020-06.k3s-test:testvol-lun-0 /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0
Output: mount: /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/10.42.2.21:3260-iqn.2020-06.k3s-test:testvol-lun-0: wrong fs type, bad option, bad superblock on /dev/sdc, missing codepage or helper program, or other error.
```

(It seems to keep retrying forever.)

This is visible in dmesg:

```
[34702.112169] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34702.198069] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34727.334700] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34727.418169] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34768.569109] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34768.659768] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34841.860618] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34841.944913] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34973.133712] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
[34973.219351] EXT4-fs (sdc): bad geometry: block count 2560256 exceeds size of device (1310720 blocks)
```

So, confirmed fixed, closing this issue, thanks! :-)

**Comment by brandond** [comment]
Glad to hear it. I spent a while tracking this one down - fortunately I was using xfs on OpenEBS so it failed rather than wiping the disk... but it was still fun to investigate.
